File: examples/python/ps7000.py
"""
Pure decoder for the Pisonics PS7000 measurement block.

Deliberately has no I/O and no dependencies: hand it a list of raw 16-bit
register values and it hands back named, scaled fields. That keeps the part
you are most likely to get wrong (word order) testable without a device on
the bench -- see test_ps7000.py.

Pisonics -- https://www.pisonics.com
"""
from __future__ import annotations

import json
import struct
from pathlib import Path

_SPEC = Path(__file__).resolve().parents[2] / "registers" / "input-registers.json"

MEASURE_BLOCK_START = 1000

# Two useful read lengths, both inside the 125-register limit of FC 0x04:
#
#   BASE  registers 1000-1086. Present on every firmware from SW 3.3.26.
#   EXT   registers 1000-1113. Adds the entrained-gas index (1104 / 1113),
#         which SW 4.x appends after the base block.
#
# Reading EXT against older firmware returns exception 02 (illegal address).
# Fall back to BASE in that case rather than treating it as a fault -- the
# measurement block is append-only, so BASE always decodes correctly.
MEASURE_BLOCK_LEN_BASE = 87
MEASURE_BLOCK_LEN_EXT = 114
MEASURE_BLOCK_LEN = MEASURE_BLOCK_LEN_BASE   # backwards-compatible default

MAIN_UNITS = {
    0: ("density", "g/cm3"),
    1: ("volume concentration", "%"),
    2: ("mass concentration", "%"),
    3: ("Baume", "degBe"),
    4: ("solids concentration", "g/L"),
    5: ("density", "kg/m3"),
}

MEASURE_ERROR_BITS = {
    0: "NOT_ENOUGH_MEMORY",
    1: "NOT_ENOUGH_DATA",
    2: "SENSOR_TEMPERATURE_INVALID",
    3: "CALIBRATION_DATA_INVALID",
    4: "CALIBRATION_FAILURE",
    5: "FLUCTUATION",
    6: "FAILED",
    7: "SOLUBILITY_CALIBRATION_DATA_INVALID",
}

# Any of these means the published value is STALE -- the firmware returned
# early this cycle and mainValueFiltered still holds the previous frame,
# while the 4-20 mA loop has already dropped to its NAMUR NE43 alarm level.
STALE_BITS = (0, 1, 3, 4, 6)


def _load_spec():
    with open(_SPEC, encoding="utf8") as fh:
        spec = json.load(fh)
    out = {}
    for block in spec["blocks"]:
        for reg in block["registers"]:
            out[reg["reg"]] = reg
    return out


def u16(regs, i):
    return regs[i] & 0xFFFF


def i16(regs, i):
    v = regs[i] & 0xFFFF
    return v - 0x10000 if v & 0x8000 else v


def u32(regs, i):
    """CDAB: low word first. This is the single most common integration bug."""
    return (regs[i] & 0xFFFF) | ((regs[i + 1] & 0xFFFF) << 16)


def f32(regs, i):
    return struct.unpack("<f", struct.pack("<I", u32(regs, i)))[0]


_READERS = {"U16": u16, "I16": i16, "U32": u32, "F32": f32}


def decode_block(regs, start=MEASURE_BLOCK_START):
    """Decode a raw register list into a dict of named, scaled values.

    regs   -- list of ints as returned by a Modbus master, index 0 == `start`
    start  -- register number of regs[0]
    """
    spec = _load_spec()
    out = {}
    for reg_no, meta in spec.items():
        idx = reg_no - start
        if idx < 0 or idx >= len(regs):
            continue
        reader = _READERS[meta["type"]]
        if idx + (2 if meta["type"] in ("U32", "F32") else 1) > len(regs):
            continue
        value = reader(regs, idx)
        if "scale" in meta:
            value = value * meta["scale"]
        out[meta["field"]] = value
    return out


def value_is_stale(measure_error_code: int) -> bool:
    """True when the instrument did not produce a new value this cycle.

    Do NOT use the `ok` field for this. `ok` reports sampling quality and is
    routinely 0 on aerated slurry while the reading is perfectly fresh; code
    that treats ok==0 as 'frozen' will silently suppress threshold alarms the
    moment the process gets noisy.
    """
    return any(measure_error_code & (1 << b) for b in STALE_BITS)


def decode_errors(measure_error_code: int):
    return [name for bit, name in MEASURE_ERROR_BITS.items()
            if measure_error_code & (1 << bit)]


def describe(decoded: dict) -> str:
    unit_code = int(decoded.get("mainValueUnit", 0))
    quantity, unit = MAIN_UNITS.get(unit_code, ("value", "?"))
    lines = [
        f"{quantity:>22}: {decoded.get('mainValueFiltered', float('nan')):.4f} {unit}",
        f"{'probe temperature':>22}: {decoded.get('sensorTemperature', float('nan')):.2f} degC"
        + ("" if decoded.get("sensorTemperatureValid") else "   (INVALID)"),
        f"{'signal quality':>22}: {decoded.get('signalQuality', float('nan')):.3f}",
        f"{'loop 1':>22}: {decoded.get('output1Current', float('nan')):.3f} mA",
        f"{'up time':>22}: {int(decoded.get('upTime', 0))} s",
    ]
    if "gasIndexEma" in decoded:
        lines.append(f"{'entrained gas (EMA)':>22}: {int(decoded['gasIndexEma'])} / 100")
    err = int(decoded.get("measureErrorCode", 0))
    if err:
        lines.append(f"{'errors':>22}: {', '.join(decode_errors(err))}")
    if value_is_stale(err):
        lines.append(f"{'!!':>22}: value is STALE, loop is at alarm level")
    return "\n".join(lines)

File: examples/csharp/Ps7000.cs
// Minimal PS7000 reader for .NET. Uses NModbus.
//
//   dotnet add package NModbus
//   dotnet add package System.IO.Ports
//   dotnet run -- COM3 1
//
// Pisonics -- https://www.pisonics.com

using System;
using System.IO.Ports;
using NModbus;
using NModbus.Serial;

namespace Pisonics
{
    /// <summary>Decoder for the PS7000 measurement block. No I/O.</summary>
    public static class Ps7000
    {
        public const ushort MeasureBlockStart = 1000;
        public const ushort MeasureBlockLenBase = 87;   // 1000-1086, all firmware
        public const ushort MeasureBlockLenExt  = 114;  // 1000-1113, SW 4.x, adds gas index

        // Registers are 1-based in the documentation; the frame carries one less.
        public static ushort FrameAddress(ushort registerNumber) => (ushort)(registerNumber - 1);

        private static int Index(ushort registerNumber) => registerNumber - MeasureBlockStart;

        public static ushort U16(ushort[] r, ushort reg) => r[Index(reg)];

        public static short I16(ushort[] r, ushort reg) => unchecked((short)r[Index(reg)]);

        /// <summary>CDAB: low word first. Getting this backwards is the
        /// number one cause of nonsense process values.</summary>
        public static uint U32(ushort[] r, ushort reg)
        {
            int i = Index(reg);
            return (uint)(r[i] | (r[i + 1] << 16));
        }

        public static float F32(ushort[] r, ushort reg) =>
            BitConverter.Int32BitsToSingle(unchecked((int)U32(r, reg)));

        // --- the fields most integrations actually need ---
        public static float MainValue(ushort[] r)      => F32(r, 1026);
        public static ushort MainValueUnit(ushort[] r) => U16(r, 1028);
        public static float ProbeTemperature(ushort[] r)  => F32(r, 1068);
        public static bool ProbeTemperatureValid(ushort[] r) => U16(r, 1067) != 0;
        public static float SignalQuality(ushort[] r)  => F32(r, 1007);
        public static float Loop1Current(ushort[] r)   => F32(r, 1045);
        public static uint  UpTimeSeconds(ushort[] r)  => U32(r, 1074);
        public static ushort MeasureErrorCode(ushort[] r) => U16(r, 1004);
        public static ushort GasIndexEma(ushort[] r)   => U16(r, 1113);

        private static readonly int[] StaleBits = { 0, 1, 3, 4, 6 };

        /// <summary>True when the instrument did not produce a new value this
        /// cycle. Do NOT use the `ok` field for this -- see docs/error-codes.md.</summary>
        public static bool ValueIsStale(ushort measureErrorCode)
        {
            foreach (int b in StaleBits)
                if ((measureErrorCode & (1 << b)) != 0) return true;
            return false;
        }

        public static string UnitName(ushort code) => code switch
        {
            0 => "g/cm3",
            1 => "% (volume)",
            2 => "% (mass)",
            3 => "degBe",
            4 => "g/L",
            5 => "kg/m3",
            _ => "?"
        };
    }

    public static class Program
    {
        public static void Main(string[] args)
        {
            string portName = args.Length > 0 ? args[0] : "COM3";
            byte slave = args.Length > 1 ? byte.Parse(args[1]) : (byte)1;

            using var port = new SerialPort(portName, 9600, Parity.None, 8, StopBits.One)
            {
                ReadTimeout = 1000,
                WriteTimeout = 1000
            };
            port.Open();

            var master = new ModbusFactory().CreateRtuMaster(port);

            ushort count = Ps7000.MeasureBlockLenExt;
            ushort[] regs;
            try
            {
                regs = master.ReadInputRegisters(
                    slave, Ps7000.FrameAddress(Ps7000.MeasureBlockStart), count);
            }
            catch (SlaveException)
            {
                // Older firmware has no gas-index registers and refuses the
                // longer read. Fall back rather than reporting a fault.
                count = Ps7000.MeasureBlockLenBase;
                regs = master.ReadInputRegisters(
                    slave, Ps7000.FrameAddress(Ps7000.MeasureBlockStart), count);
            }

            ushort err = Ps7000.MeasureErrorCode(regs);

            Console.WriteLine($"     main value : {Ps7000.MainValue(regs):F4} " +
                              Ps7000.UnitName(Ps7000.MainValueUnit(regs)));
            Console.WriteLine($"    probe temp. : {Ps7000.ProbeTemperature(regs):F2} degC" +
                              (Ps7000.ProbeTemperatureValid(regs) ? "" : "   (INVALID)"));
            Console.WriteLine($" signal quality : {Ps7000.SignalQuality(regs):F3}");
            Console.WriteLine($"         loop 1 : {Ps7000.Loop1Current(regs):F3} mA");
            Console.WriteLine($"        up time : {Ps7000.UpTimeSeconds(regs)} s");
            if (count == Ps7000.MeasureBlockLenExt)
                Console.WriteLine($"  entrained gas : {Ps7000.GasIndexEma(regs)} / 100");
            if (Ps7000.ValueIsStale(err))
                Console.WriteLine($"             !! : value is STALE (measureErrorCode 0x{err:X4})");
        }
    }
}
